### PR TITLE
sys/shell: adapt ifconfig to work with sx1272/76 devices

### PR DESCRIPTION
**Pull request by aabadie**
As required in #7356, I adapted the ifconfig code to work with sx127x devices as well. Finally, not so difficult, but this code is becoming quite difficult to follow with all those #ifdefs.

Based on ~#7356~ ~#8160~ for the netif/auto_init adaption.

### CODE REVIEW

**Comment by smlng** [comment]
I see a problem with the proposed implementation here, at least the `usage` help text is **either** for LoRa with SX127x **or** for any other network device, also the command description for `ifconfig` is changed.

**Comment by aabadie** [comment]
I'm not sure to understand your comment.
The implementation of `_usage` is conditional (using macros) in regard to the type of device which is normal since LoRa devices (SX127x) are quite different from other devices, thus the usage message is different.
For the moment, only SX127x devices are supported for LoRa, this may change in the future if we provide support for other LoRa devices.

**Comment by miri64** [comment]
What @smlng is saying is basically, what I was saying in #7356. It doesn't make sense, that LoRA gets a special command for interface configuration, since it is also just a network device / interface.

**Comment by aabadie** [comment]
Just pushed an update, let me know if it's better

**Comment by smlng** [comment]
> since it is also just a network device / interface.

I'm not sure if it is that simple, but in general I agree. 

However, judging from @aabadie s implementation it looks like that a LoRa interface has several additional attributes/parameters compared to ethernet or 15.4 interface while other are even missing or are not supported. 

Maybe it does need an additional command to configure those special attributes, similar to `iwpan` on Linux for 15.4 or `iwconfig` for wifi a while back.

TL;DR this may need some thinking  as the `ifconfig` command is getting quiet complex if it has to support diverse radios with loads of different and special parameters.

**Comment by aabadie** [comment]
> Maybe it does need an additional command to configure those special attributes, similar to iwpan on Linux for 15.4 or iwconfig for wifi a while back.

What about a `lora` generic command ? It could have several subcommands:
* `lora config`: list available lora interfaces, get/set parameters
* `lora send <if> <data>`: send raw radio paquets
* `lora listen <if>`: listen to raw radio incoming paquets
* `lora mac <join|tx|set|get>`: configure the LoRamac layer (not available yet)

The get/set functions (and maybe others) could be shared between the actual `ifconfig` code and the `lora` code in a separate `.c` file.

**Comment by miri64** [comment]
> lora send <if> <data>: send raw radio paquet

Why is that one needed? What is different with lora, that `txtsnd` doesn't work?

> lora listen <if>: listen to raw radio incoming paquets

Same goes here.

**Comment by aabadie** [comment]
> Why is that one needed? What is different with lora, that txtsnd doesn't work?

I didn't first of it initially and I just pushed an update with this change.

For the lora listen part, that's indeed true that it's not required since one can achieve the same result by putting the radio in RX state with ifconfig.

**Comment by aabadie** [comment]
@miri64, @smlng, do you other comments on this one ?

**Comment by aabadie** [comment]
@miri64, I would like to adapt this PR to #8159 when it's merged but before I want to be sure that it makes sense to modify the ifconfig/txtsnd commands to work with LoRa radio.
The ifconfig code also has changed after the netif remodeling so adapting it will probably end up in a lot of `#ifdef`.

**Comment by miri64** [comment]
> @miri64, I would like to adapt this PR to #8159 when it's merged but before I want to be sure that it makes sense to modify the ifconfig/txtsnd commands to work with LoRa radio.

When it comes to new netopts to be supported (IQ_INVERTED etc.) it makes sense. Yes!

The ifconfig code also has changed after the netif remodeling so adapting it will probably end up in a lot of #ifdef.

No, they don't require any `#ifdef`s or `#ifndef`s. If the option is not supported the network device returns `-ENOTSUP` so the command catches this. You don't see any `#ifdef`s for Ethernet or IEEE 802.15.4 in this here either, though they are very different protocols, right?

**Comment by miri64** [comment]
Needs also adaptation for #7925.

**Comment by aabadie** [comment]
I reworked this PR to include the recent refactoring of netif.

Now this PR is based on #8160. I ran a few tests with another radio using the driver tests application and could exchange some packets between the 2 (one using ifconfig from default application and one using the test application).

File: sys/shell/commands/sc_gnrc_netif.c [hunk 1]
```diff
@@ -31,6 +31,9 @@
 #ifdef MODULE_L2FILTER
 #include "net/l2filter.h"
 #endif
+#ifdef MODULE_SX127X
+#include "net/lora.h"
+#endif
```
Comment (miri64) [inline sys/shell/commands/sc_gnrc_netif.c]: Please remove all those `#ifdef`s and `#ifndef`s. They are not necessary. Especially the `#ifndefs` shouldn't be there. And if you want to keep the `#ifdef`s, please use another module (e.g. [a pseudo-module] `netdev_lora`).

Comment (aabadie) [reply sys/shell/commands/sc_gnrc_netif.c]: I know I can remove them but if I do that I have 2 problems:
* unrelated options/parameters are displayed in the usage output
* the size of firmware increases for other boards (e.g samr21-xpro)

I'll change that for the pseudo module.

Comment (miri64) [reply sys/shell/commands/sc_gnrc_netif.c]: > unrelated options/parameters are displayed in the usage output

a) That wasn't a problem for other types of devices. A user who has only an Ethernet device also sees the options to set a channel, but will notice it was stupid to try to set them as soon as they see the error message `ifconfig` reports, because the device does not support them.
b) You exclude the IPv6-options. IPv6 over LoRA is [in the works](https://datatracker.ietf.org/wg/lpwan/about/), so removing them seems a little short-sighted.
c) If a board has two interfaces of which only one board is a LoRa board, the user wouldn't be able to configure the other interface properly, because the LoRA device is compiled in. This is a valid scenario, since e.g. LoRA-to-IEEE802.15.4 border routers is something we already sketched out.

> the size of firmware increases for other boards (e.g samr21-xpro)

That is a valid argument, but only if you make it dependent on LoRA, not on the device.

Comment (aabadie) [reply sys/shell/commands/sc_gnrc_netif.c]: Added the pseudo module and removed a few ifdefs.

File: sys/shell/commands/sc_gnrc_netif.c [hunk 2]
```diff
@@ -360,10 +398,18 @@ static void _netif_list(kernel_pid_t iface)
         printf(" HWaddr: %s ",
                gnrc_netif_addr_to_str(hwaddr, res, hwaddr_str));
     }
+#ifdef MODULE_LORA
+    uint32_t u32;
+    res = gnrc_netapi_get(iface, NETOPT_CHANNEL, 0, &u32, sizeof(u32));
+    if (res >= 0) {
+        printf(" Channel: %" PRIu32 " ", u32);
+    }
```
Comment (miri64) [inline sys/shell/commands/sc_gnrc_netif.c]: Soooo... you can't set the channel of an IEEE 802.15.4 device with LoRA anymore now? Because for all the devices that have a 16-bit wide channel as documented in `netopt.h` this will evaluate to something random...

File: sys/shell/commands/sc_gnrc_netif.c [hunk 3]
```diff
@@ -514,6 +589,118 @@ static void _netif_list(kernel_pid_t iface)
...
+
+    if (_is_number(u32_str)) {
+        if ((res = strtoul(u32_str, NULL, 10)) == ULONG_MAX) {
+            puts("error: unable to parse value.\n"
+                 "Must be a 32-bit unsigned integer (dec or hex)\n");
+            return 1;
+        }
+    }
+    else {
+        if ((res = strtoul(u32_str, NULL, 32)) == ULONG_MAX) {
```
Comment (miri64) [inline sys/shell/commands/sc_gnrc_netif.c]: I IIRC `strtoul()` isn't portable.

File: sys/shell/commands/sc_gnrc_netif.c [hunk 4]
```diff
@@ -805,9 +992,24 @@ static int _netif_set(char *cmd_name, kernel_pid_t iface, char *key, char *value
     else if (strcmp("cca_threshold", key) == 0) {
         return _netif_set_u8(iface, NETOPT_CCA_THRESHOLD, 0, value);
     }
+#ifdef MODULE_LORA
+    else if ((strcmp("channel", key) == 0) || (strcmp("chan", key) == 0)) {
+        return _netif_set_u32(iface, NETOPT_CHANNEL, 0, value);
```
Comment (miri64) [inline sys/shell/commands/sc_gnrc_netif.c]: Channel problem again.

**Comment by miri64** [comment]
Needs rebase.

File: sys/shell/commands/sc_gnrc_netif.c (hunk 2 above)
Comment (aabadie) [reply sys/shell/commands/sc_gnrc_netif.c]: So what is the best solution ? Adding another NETOPT_LORA_CHANNEL ? or just use u32 for all devices ?

Comment (aabadie) [reply sys/shell/commands/sc_gnrc_netif.c]: There is the same issue with TX power: SX127x defines an int8_t whereas other radios have int16_t. But it will be easy to use int16_t for the sx127x as well.

File: sys/shell/commands/sc_gnrc_netif.c (hunk 3 above)
Comment (aabadie) [reply sys/shell/commands/sc_gnrc_netif.c]: Maybe but then the shell commands are already not portable, see [here](https://github.com/RIOT-OS/RIOT/blob/master/sys/shell/commands/sc_gnrc_netif.c#L524) (In fact I just copy pasted this function and replace uint16_t with uint32_t...)

Comment (miri64) [reply sys/shell/commands/sc_gnrc_netif.c]: Alright.

File: sys/shell/commands/sc_gnrc_netif.c (hunk 2 above)
Comment (miri64) [reply sys/shell/commands/sc_gnrc_netif.c]: > So what is the best solution ? Adding another NETOPT_LORA_CHANNEL ? or just use u32 for all devices ?

Why is this value `uint32_t` anyway? Are there really LoRA channels >65535? This is where we should start our discussion on the requirement of an API change ;-).

Comment (miri64) [reply sys/shell/commands/sc_gnrc_netif.c]: > There is the same issue with TX power: SX127x defines an int8_t whereas other radios have int16_t. But it will be easy to use int16_t for the sx127x as well.

Why is this an issue? Just cast the value to `int8_t` and make sure (with an assert) that the provided value is within the range (`INT8_MIN`, `INT8_MAX`) (both things should of course be happening either in the device driver or within a #7736-style LoRA-layer).

Comment (aabadie) [reply sys/shell/commands/sc_gnrc_netif.c]: > Why is this value uint32_t anyway?

LoRa channels correspond to the center frequency in Hz, e.g 868100000, 868300000 in Europe. That's why uint32_t is required. Maybe we can use some enums and stick to the uint16_t but then it becomes more complex in the driver implementation because regions have different channels values (915000000 for US, etc) and number.

Comment (bergzand) [reply sys/shell/commands/sc_gnrc_netif.c]: Would it work to define this value as the center frequency multiplied by 100Khz? Then the values would be between 8681, 8683 etc?

Comment (miri64) [reply sys/shell/commands/sc_gnrc_netif.c]: Ok, then since this is different enough from other radio technologies, I would rather vote to introduce another NETOPT value. But not `LORA_CHANNEL`. How about `NETOPT_FREQ_BAND` for (center of a) frequency band?

Comment (miri64) [reply sys/shell/commands/sc_gnrc_netif.c]: (this way it can also be used by other technologies that use a similar channel scheme)

Comment (aabadie) [reply sys/shell/commands/sc_gnrc_netif.c]: > How about NETOPT_FREQ_BAND 

+1, I already started to do this change ;)

Comment (aabadie) [reply sys/shell/commands/sc_gnrc_netif.c]: I finally opted for NETOPT_CHANNEL_FREQUENCY which is more IMHO in the spirit of LoRa technology. The band is more for 868MHz, 915MHz, 433MHz, etc. Each band is divided into channels of different widths (125kHz, 250kHz and 500kHz)

Comment (miri64) [reply sys/shell/commands/sc_gnrc_netif.c]: @bergzand

> Would it work to define this value as the center frequency multiplied by 100Khz? Then the values would be between 8681, 8683 etc?

I think this would be more confusing than having just a separate option.

Comment (bergzand) [reply sys/shell/commands/sc_gnrc_netif.c]: > I think this would be more confusing than having just a separate option.

Yeah, I was trying to fit it into the existing netopt here, I agree with introducing a new netopt value for this, probably more clear than trying to multiplex different interpretatioins over a single netopt. :)

File: sys/include/net/netopt.h [hunk 5]
```diff
@@ -375,6 +375,11 @@ typedef enum {
      */
     NETOPT_INTEGRITY_CHECK,
 
+    /**
+     * @brief   Get/Set the channel frequency as uint32_t.
```
Comment (miri64) [inline sys/include/net/netopt.h]: Maybe add some detailed description (i.e. paragraph under this line), about how this differs from the `NETOPT_CHANNEL` option.

Comment (miri64) [reply sys/include/net/netopt.h]: Ping @aabadie?

Comment (aabadie) [reply sys/include/net/netopt.h]: Not sure if better, but I added the detailed description.

Comment (miri64) [reply sys/include/net/netopt.h]: Maybe not perfect, but sufficient for me.

File: sys/shell/commands/sc_gnrc_netif.c [hunk 6]
```diff
@@ -372,6 +422,20 @@ static void _netif_list(kernel_pid_t iface)
     if (res >= 0) {
         printf(" NID: 0x%" PRIx16, u16);
     }
+#ifdef MODULE_LORA
```
Comment (miri64) [inline sys/shell/commands/sc_gnrc_netif.c]: You removed the other `#ifdef`s. Why not this one (which except for size optimization is the least useful as it wouldn't show anything if compiled in but the options were not supported)?

File: sys/shell/commands/sc_gnrc_netif.c [hunk 7]
```diff
@@ -805,6 +987,20 @@ static int _netif_set(char *cmd_name, kernel_pid_t iface, char *key, char *value
     else if (strcmp("cca_threshold", key) == 0) {
         return _netif_set_u8(iface, NETOPT_CCA_THRESHOLD, 0, value);
     }
+    else if ((strcmp("frequency", key) == 0) || (strcmp("freq", key) == 0)) {
+        return _netif_set_u32(iface, NETOPT_CHANNEL_FREQUENCY, 0, value);
+    }
+#ifdef MODULE_LORA
```
Comment (miri64) [inline sys/shell/commands/sc_gnrc_netif.c]: Dito. Usage above isn't `#ifdef`'d

Comment (aabadie) [reply sys/shell/commands/sc_gnrc_netif.c]: @miri64 I pushed 4ccfcfd that removes the ifdefs.

**Comment by aabadie** [comment]
@miri64 I guess we are good here, shall I squash ?

File: sys/shell/commands/sc_gnrc_netif.c [hunk 8]
```diff
@@ -514,6 +578,116 @@ static void _netif_list(kernel_pid_t iface)
...
+        if ((res = strtoul(u32_str, NULL, 32)) == ULONG_MAX) {
+            puts("error: unable to parse value.\n"
+                 "Must be a 32-bit unsigned integer (dec or hex)\n");
+            return 1;
+        }
+
+        hex = true;
+    }
+
+    if (res > 0xffffffff) {
```
Comment (miri64) [inline sys/shell/commands/sc_gnrc_netif.c]: `UINT32_MAX`? Also, I think it is safe to make this an assert for now. I don't know of any platform in our scope where `sizeof(unsigned long) > sizeof(uint32_t)`.

Comment (aabadie) [reply sys/shell/commands/sc_gnrc_netif.c]: Done

**Comment by aabadie** [comment]
@miri64 shall I squash this one ?

**Comment by aabadie** [comment]
squashed !

**Comment by aabadie** [comment]
@miri64, I have overflows issues with Murdock. Is it ok if I add some ifdefs to skip the LoRa specific functions ?

**Comment by miri64** [comment]
No, just put the `z1` into `BOARD_INSUFFICIENT_MEMORY`.

**Comment by miri64** [comment]
But we definitely need a modular approach to ifconfig in the future. The current approach doesn't scale (as would `#ifdef` in such a case, see `git log sys/transceiver/` to see the horrors that let us hate `#ifdef`s ;-)). Alternatively, extra commands to configure IEEE 802.15.4-, Ethernet-, LoRA-, ...-specific options separately as @jfischer-phytec-iot once proposed could help.

**Comment by miri64** [comment]
(last proposal would be different than the original version of this PR would be that interface configuration from network layer view would still be done with `ifconfig`, regardless of the link-layer technology [compare iwconfig and ifconfig in Linux).
